sklearn/preprocessing/_encoders.py

```diff
@@ -509,7 +510,7 @@ def inverse_transform(self, X):
         found_unknown = {}
 
         for i in range(n_features):
-            if self.drop is None:
+            if (self.drop is None) or (self.drop_idx_[i] is None):
```

Comment:
I don't think that we need the parenthesis

```suggestion
            if self.drop is None or self.drop_idx_[i] is None:
```

I think that we could change to only use the fitted attribute

```python
if self.drop_idx_ is None or self.drop_idx_[i] is None:
```

Reply:
I might expect a bug if we are going to reset drop since `drop` and `drop_idx_` could be in different fitting stage.

```python
ohe = OneHotEncoder(drop="if_binary")
ohe.fit(...)
ohe.set_params(drop=None)
ohe.get_features_names()
```
we could check the example above would lead to something fishy. It is just on the top of the head.

Reply:
If it does, we would need to add a non-regression test :)

Reply:
Sorry @glemaitre this comment about the fitted attribute is not clear to me. I haven't change anything for now. Running the following code
```
from sklearn.datasets import fetch_openml
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split

survey = fetch_openml(data_id=534, as_frame=True)
X = survey.data[survey.feature_names]
y = survey.target.values.ravel()

X_train, X_test, y_train, y_test = train_test_split(
    X, y, random_state=42
)

categorical_columns = ['RACE', 'OCCUPATION', 'SECTOR',
                       'MARR', 'UNION', 'SEX', 'SOUTH']

ohe = OneHotEncoder(drop="if_binary")
ohe.fit(X_train[categorical_columns])
print(ohe.get_feature_names())
ohe.set_params(drop=None)
print(ohe.get_feature_names())
```
gives consistent results, IMO.
Is that what you are expecting?

Reply:
Try the following:

```python
from sklearn.datasets import fetch_openml
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split

survey = fetch_openml(data_id=534, as_frame=True)
X = survey.data[survey.feature_names]
y = survey.target.values.ravel()

X_train, X_test, y_train, y_test = train_test_split(
    X, y, random_state=42
)

categorical_columns = ['RACE', 'OCCUPATION', 'SECTOR',
                       'MARR', 'UNION', 'SEX', 'SOUTH']

ohe = OneHotEncoder(drop="if_binary")
X_trans = ohe.fit_transform(X_train[categorical_columns])
print(X_train[categorical_columns].shape)
print(ohe.inverse_transform(X_trans).shape)
ohe.set_params(drop=None)
print(ohe.inverse_transform(X_trans).shape)
```

Reply:
```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/tmp/tmp.py in <module>
     19 print(ohe.inverse_transform(X_trans).shape)
     20 ohe.set_params(drop=None)
---> 21 print(ohe.inverse_transform(X_trans).shape)

~/Documents/packages/scikit-learn/sklearn/preprocessing/_encoders.py in inverse_transform(self, X)
    501                "columns, got {1}.")
    502         if X.shape[1] != n_transformed_features:
--> 503             raise ValueError(msg.format(n_transformed_features, X.shape[1]))
    504 
    505         # create resulting array of appropriate dtype

ValueError: Shape of the passed X data is not correct. Expected 20 columns, got 16.
```

Reply:
If we only use the fitted attributes then we will not get this issue.